sgqlc/endpoint/websocket.py

```diff
@@ -1,14 +1,73 @@
+from typing import Any, Dict, Union, SupportsBytes
+from websocket import WebSocket
+
 from sgqlc.endpoint.base import BaseEndpoint
 import websocket
 import uuid
 import json
 
 
+def _receive_message(ws: WebSocket) -> Dict[Any, Any]:
+    '''
+    Receive and return  message from the websocket,
+    discarding any keepalive messages
+
+    :param ws: websocket to listen on
+    :type ws: WebSocket
+
+    :return: message from the server
+    :rtype: dict
+    '''
+    response = {'type': 'ka'}
+    while response['type'] == 'ka':
+        response = json.loads(ws.recv())
+    return response
+
+
+class CancelableResultGenerator:
+    '''
+    Generator of websocket subscription notification events that supports
+    sending a message to the server to cancel the subscription.
+    '''
+    def __init__(self, ws: WebSocket, query_id: str):
+        '''
+        :param ws: websocket to get results from
+        :type ws: WebSocket
+
+        :param query_id: query_id to track
+        :type query_id: str
+        '''
+        self.ws = ws
+        self.query_id = query_id
+
+    def __next__(self) -> dict:
+        response = _receive_message(self.ws)
+        if response['id'] != self.query_id:
+            raise ValueError(
+                f'Unexpected id {response["id"]} '
+                f'when waiting for query results'
+            )
+        if response['type'] == 'data':
+            return response['payload']
```

Comment:
this 2 `if` statements is what I think we'll need to change to be something like a dispatcher.

`CancelableResultGenerator`  I'd rename to `WebSocketHandler`, it would have an API such as:

```py
class WebSocketHandler:
   _id_handlers: Dict[str, Callable[[Dict[str, Any]], None]]
   
   def register_handler(self, _id: str, handler: Callable[[Dict[str, Any]], None]) -> None:
      self._id_handlers[_id] = handler

  def unregister_handler(self, _id: str):
     del self._id_handlers[_id]

    def __next__(self) -> None:  # it won't emit values on its own
        response = _receive_message(self.ws)  # maybe rewrite as self._receive_reponse() and would also do "connection_ack" if not done yet
        if response['type'] == 'data':
          handler = self._id_handlers[response['id']]
          handler(response)

        ... the rest of your code here...


class WebSocketEndpoint:
    def __call__(self, ...):
       # very rough here, would need to think exactly how to do this properly:
       self._pending_response = None
       self.ws_handler.register_handler(query_id, self._handle_response)
       ...
       try:
           while self.ws_handler.__next__():
               response = self._pending_response
               if response is not None:  # it may be another query that was handled
                  self._pending_response = None
                  yield response['payload']
       finally:
           if self._created_ws_handler:
              self.ws_handler.close() # only if we created it  

   def _handle_response(self, response):
       self._pending_response = response
```

It's very rough, maybe i missed something but I think this would keep the same semantics as in your usage. But it allows things such as a SubscriptionEndpoint to callback:
```py
class WebSocketAsyncEndpoint:
    def __call__(self, ..., on_data):
      def handle_response(response):
          on_data(response['payload'])
      self.ws_handler.register_handler(query_id, self._handle_response)
      # does NOT iterate
      def unregister():
         self.ws_handler.unregister_handler(query_id)
      return unregister
```

Then some central place would iterate and multiple async would co-exist.